flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/TimerService.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks;
+
+import org.apache.flink.annotation.Internal;
+
+/**
+ *
+ * <p>The registration of timers follows a life cycle of three phases:
+ * <ol>
+ *     <li>In the initial state, it accepts timer registrations and triggers when the time is reached.</li>
+ *     <li>After calling {@link #quiesce()}, further calls to
+ *         {@link #registerTimer(long, ProcessingTimeCallback)} will not register any further timers, and will
+ *         return a "dummy" future as a result. This is used for clean shutdown, where currently firing
+ *         timers are waited for and no future timers can be scheduled, without causing hard exceptions.</li>
+ *     <li>After a call to {@link #shutdownService()}, all calls to {@link #registerTimer(long, ProcessingTimeCallback)}
+ *         will result in a hard exception.</li>
+ * </ol>
+ */
+@Internal
+public interface TimerService extends ProcessingTimeService {
```

Comment:
Although, the commit message says that the lifecycle methods were extracted into a separate interface, my main motivation was to separate "application level api" (`ProcessingTimeService` that exposed to operators) from the system timer provider (implementation of `TimerService`).
It just happens that currently in this PR `TimerService` piggybacks on `ProcessingTimeService`
interface (and `ProcessingTimeCallback`) to define the system timer api.

I can introduce a new, separate timer api for `TimerService` to explicitly separate them. It should not make big changes in the runtime code, only some tests should be affected to be more explicit what they are testing against. I'd like to hear from others if you are ok with this direction.

Regarding the naming, `ProcessingTimeService` is simply the preserved name that operators see and `TimerService` is a common name in Flink code base for classes with similar functionality.

Introducing a common "life-cycle" interface may be too early generalization on my side.
Also, unifying/consolidating the `TimerService` classes may be a separate effort on it's own.

Reply:
I'm not sure if naming this interface `TimerService` is a good idea. Having 4 different interfaces/classes with the same name is pretty confusing. I would prefer to prefix them like `CEPTimerService`.

@AHeise. Regarding the extraction of the lifecycle methods to a `Lifecycle` interface and implementing it in the multiple independent places. If those different classes are are not inter-exchangeably used, so the common base `Lifecycle` interface is used only to make sure that the naming convention remains consistent, I would be strongly against that. Recently I've introduce this with the `AsyncDataInput` that was implemented by couple of independent classes, and honestly that was a mistake. Whenever I was trying to look for usages of the `SingleInputGate#pollNext()` method, IntelliJ was pointing me to tons of the unrelated calls to for example `StreamTaskInput#pollNext()`. (Currently the situation has improved in the code as we introduced `PullingAsyncDataInput` and `PushingAsyncDataInput`.

Reply:
Extracting Lifecycle methods is purely optional. @pnowojski you can make it rather small like https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/context/Lifecycle.html or https://logging.apache.org/log4j/2.x/log4j-core/apidocs/org/apache/logging/log4j/core/LifeCycle.html and add specific stuff (`quiensce`) only to specific implementations. But like @1u0 said, it may be premature at this point in time. If we manage to pull life-cycle methods from user-facing interfaces, we can do that later.

Anyways, I'd still like to address the naming. My main source of confusion is the difference between `TimeService` and `TimerService`.

Reply:
Okay @1u0 convinced my that `ProcessingTime` is well-known name on its own, so that `ProcessingTimeService` makes sense. Similarly, `TimerService` on its own makes sense.
So even though it seems inconsistent, I would keep it as is in this PR.